### PR TITLE
 Modify the build_optimizer method and add the config parameter

### PR DESCRIPTION
**Pull request by blue-q**
I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Introducing customizable learning rate fine-tuning and optimizer adjustments! 🚀

### 📊 Key Changes
- **New Learning Rate Settings**: Added options to fine-tune learning rates for specific layers or sets of layers.
- **Enhanced Optimizer Creation**: Updated the optimizer building process to support the new learning rate configurations.
- **Flexibility in Training**: Users can now specify groups of parameters that have different learning rates, improving model performance by focusing on the optimization of crucial layers.

### 🎯 Purpose & Impact
- **Customized Training**: The introduction of parameter groups allows for more nuanced control over the training process, potentially leading to enhanced model accuracy.
- **Improved Efficiency**: By adjusting the learning rates for certain layers, training can become more efficient, speeding up the learning process or improving results without increasing training time.
- **Versatility for Users**: Both novices and experts can benefit from easier customization of their model training process, making it simpler to tailor the learning to specific needs or experiments.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR introduces fine-tuning settings for learning rates that allow for more granular control over optimizer behavior in model training.

### 📊 Key Changes
- Added new configuration options for learning rate fine-tuning in `default.yaml`.
  - `lr_group`: Enables grouped learning rates.
  - `scaling_ratio`: Defines the scaling ratio for learning rates.
  - `scaling_layer`: Specifies which layers should have scaled learning rates.
- Enhanced `build_optimizer` function in `trainer.py` to support the new learning rate fine-tuning configurations, allowing different learning rates for different groups of model parameters.

### 🎯 Purpose & Impact
- **Enhanced Customization**: Users can now fine-tune learning rates for specific groups of layers, which can help optimize training performance.
- **Improved Training Optimization**: With finer control over learning rates, models can potentially converge faster and achieve better accuracy.
- **Flexibility for Advanced Users**: These changes offer more flexibility for users who need advanced training customizations, potentially leading to improved model performance. 🌟

Overall, these changes are designed to provide users with greater control over the training process, leading to potentially improved results. 🚀

### CODE REVIEW

**Comment by blue-q** [comment]
recheck

**Comment by blue-q** [comment]
recheck

**Comment by blue-q** [comment]
#9836

**Comment by blue-q** [comment]
recheck

**Comment by glenn-jocher** [comment]
It's great to have you as part of the community and thank you for signing the CLA! 🙌

For your PR on customizable learning rates and optimizer adjustments, it sounds like a valuable enhancement! 🚀 Could you please ensure that any new parameters introduced are well-documented in the method's docstring? Also, including a simple example in your PR description demonstrating the use of these new configurations can significantly help others grasp the benefits more quickly.

For instance, if your changes allow users to specify different learning rates for different layers, a quick snippet like this could be illustrative:

```python
optimizer_config = {
    "lr_groups": {
        "backbone": 0.001,
        "head": 0.01
    }
}
model.build_optimizer(config=optimizer_config)
```
This simple example can demonstrate the practical application of your enhancements. Keep up the great work, and I look forward to seeing your PR progress!

**Comment by glenn-jocher** [comment]
@blue-q hi there! 😊 It sounds like you're making great progress with your PR. Adjusting learning rates for specific layers is definitely a powerful feature. Thanks for sharing your approach and example; it clarifies how to use the new parameters you introduced.

For using `scaling_layer` and `scaling_ratio`, your example is spot on. Just a small tip: when sharing code or new features like this, you may also want to highlight how users can revert to default behaviors or adjust these settings for different scenarios. This can be as simple as mentioning:
```python
# To apply default learning rates to all layers
model.train(data='coco128.yaml', epochs=100, imgsz=640)

# For customizing learning rates through `default.yaml`
# layers_lr: {22: 0.01, "0:21": 0.001}
```
It's great to see contributions like yours that enhance flexibility in training configurations. Keep up the amazing work! If you need further assistance or have more ideas, feel free to reach out. 👍

**Comment by blue-q** [comment]
your suggestion is excellent, so I have added an example to illustrate it. 
```
# To apply default learning rates to all layers
model.train(data='coco128.yaml', epochs=100, imgsz=640)
```

**Comment by blue-q** [comment]
Hi @glenn-jocher , your suggestion is excellent, so I have added an example to illustrate it. 
```
# To apply default learning rates to all layers
model.train(data='coco128.yaml', epochs=100, imgsz=640)

#To apply different learning rates for different layers
model.train(data='coco128.yaml', epochs=100, imgsz=640, param_group=True, scaling_layer=22, scaling_ratio=10.0) #layers_lr: {22: lr0, "0:21": lr0/10}
```
Of course, you can do this directly by modifying 'default.yaml'.

**Comment by glenn-jocher** [comment]
Hi @blue-q! 🌟 Your addition looks spot on and the `lr_group` naming indeed aligns better with its functionality. It's clear and straightforward—the code examples are a great touch for understanding how to use different learning rates. Thanks for making that change and adding such a helpful example! This enhancement to customizable learning rates is bound to be a fantastic feature for users looking to fine-tune their training process. Great work! 👍

**Comment by blue-q** [comment]
@glenn-jocher now it looks like the merge of code seems to have some problems, now how do I deal with it

**Comment by glenn-jocher** [comment]
Hi @blue-q! 👋 It seems there might be a merge conflict. Here’s a quick way to resolve it:

1. **Update Your Branch**: Ensure your branch is updated with the latest main branch changes. You can do this by running:
   ```
   git fetch origin
   git rebase origin/main
   ```
2. **Resolve Conflicts**: Check for conflicts and manually resolve them in your code editor.
3. **Continue Rebase**: Once conflicts are resolved, continue the rebase process with `git rebase --continue`.
4. **Push Changes**: Push your changes with `git push --force` to update the PR.

If you need more detailed guidance, feel free to ask! 😊

**Comment by blue-q** [comment]
Hi @glenn-jocher , I clicked `update with rebase` this morning, and now All checks have passed, but it needs at least 1. review is required by reviewers with write access, and showing` Merging is blocked  Merging can be performed automatically with 1 approving review.`

**Comment by blue-q** [comment]
hi @glenn-jocher,  can I please @ someone, or can you help me @ them

**Comment by glenn-jocher** [comment]
Hi @blue-q! 👋

Absolutely, you can @mention someone to get their attention. If you need a specific reviewer or maintainer to look at your PR, you can tag them directly in your comment like this:

```markdown
@maintainer-username Could you please review this PR when you have a moment? Thanks! 😊
```

If you're unsure who to tag, you can check the repository's contributors or maintainers list. If you still need assistance, feel free to let me know, and I'll help guide you to the right person.

Thanks for your patience and contributions! 🚀

**Comment by blue-q** [comment]
Hi @Burhan-Q ，could you please review this PR when you have a moment? Thanks! 😊

**Comment by blue-q** [comment]
Hi @glenn-jocher ，can you help me find someone to review my PR.

**Comment by glenn-jocher** [comment]
Hi @blue-q,

Thank you for your patience. I've noticed your request for a review. Please allow some time for the maintainers to get to it. If there's no response soon, you might consider tagging another active contributor or maintainer. Your contributions are appreciated!

**Comment by blue-q** [comment]
Hi @glenn-jocher ，can you help me find someone to review my PR. This merging is blocked.(Merging can be performed automatically with 1 approving review.)

**Comment by glenn-jocher** [comment]
Hi @blue-q,

Thank you for your patience. I've noted your request for a review. Please allow some time for the maintainers to address it. If there's no response soon, you might consider tagging another active contributor or maintainer. Your contributions are appreciated!

**Comment by blue-q** [comment]
Hi，@Burhan-Q can you help me find someone to review my PR.

**Comment by blue-q** [comment]
Hi @ambitious-octopus  can you help me find someone to review my PR.
